Database code refactoring.

ashwathraop opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
Code refactor for database related code within PBSPro.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
For easy maintainability, current database-related code within the PBS server needs to be refactored into a separate pluggable dynamic library. Another motivation behind this refactoring is the added advantage of moving to any database without having to do major code changes within PBS itself.

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->
https://pbspro.atlassian.net/wiki/spaces/PD/pages/1524563969/DB+Refactor+Design

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->

<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
Tests in progress.


<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

hirenvadalia commented on pbspro.spec:
```diff
@@ -477,15 +480,13 @@ ${RPM_INSTALL_PREFIX:=%{pbs_prefix}}/libexec/pbs_posttrans \
 %exclude %{pbs_prefix}/lib*/init.d/sgiICEplacement.sh
 %exclude %{pbs_prefix}/lib*/python/altair/pbs_hooks/*
-%exclude %{pbs_prefix}/libexec/install_db
+%exclude %{pbs_prefix}/libexec/pbs_db_utility.sh
```
Do we really need `.sh` extension here?

ashwathraop replied:
Its a shell script. I know we are not following a standard. But I feel better having it.

ashwathraop replied:
We dont have it as a standard. We have scripts with and without .sh extension. Since most of the files I touched dont have .sh in them, I removed the extension.

hirenvadalia commented on src/cmds/Makefile.am:
```diff
@@ -118,11 +119,24 @@ pbs_demux_CPPFLAGS = ${common_cflags}
+pbs_dataservice_bin_LDFLAGS = "-Wl,-rpath,$(DESTDIR)$(libdir)"
+pbs_dataservice_bin_LDADD = \
+    $(top_builddir)/src/lib/Libdb/libdb.so \
+	${common_libs} \
```
needs indentation fix

hirenvadalia commented on src/cmds/pbs_dataservice.c:
```diff
@@ -0,0 +1,138 @@
+ * trademark licensing policies.
+ *
+ */
+
```
remove extra empty lines

hirenvadalia commented on src/cmds/pbs_dataservice.c:
```diff
@@ -0,0 +1,138 @@
+	int rc;
+	int i;
+	int errflg = 0;
+	char prog[]="pbs_dataservice";
```
space around =

hirenvadalia commented on src/cmds/pbs_dataservice.c:
```diff
@@ -0,0 +1,138 @@
+	char conn_db_host[PBS_MAXSERVERNAME+1];
+	char *errmsg = NULL;
+
+	/*test for real deal or just version and exit*/
```
space around *

hirenvadalia commented on src/cmds/pbs_dataservice.c:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	/* check admin privileges */
+	if ((getuid() != 0) || (geteuid() != 0)) {
```
remove redundant braces

hirenvadalia commented on src/cmds/scripts/pbs_dataservice:
```diff
@@ -212,136 +197,31 @@ check_status() {
-# 
-
 case "$1" in
 	start|startasync)
```
I think now we don't "startasync" anymore right?

hirenvadalia commented on src/cmds/scripts/pbs_ds_password:
```diff
@@ -39,7 +39,6 @@
-exec $PBS_EXEC/sbin/pbs_ds_password.bin ${1+"$@"}
+. "$PBS_EXEC"/libexec/pbs_db_env.sh
 
+exec $PBS_EXEC/sbin/pbs_ds_password.bin ${1+"$@"}
```
missing new line

hirenvadalia commented on src/cmds/scripts/pbs_habitat.in:
```diff
@@ -635,11 +283,11 @@ if [ "${PBS_START_SERVER:-0}" != 0 ] ; then
 		exit 1
 	fi
-	resp=`${PBS_EXEC}/libexec/install_db 2>&1`
+	resp=`${PBS_EXEC}/libexec/pbs_db_utility.sh "install_db" 2>&1`
```
do we really need quote in the argument?

ashwathraop replied:
Yes. some flavors of Linux are very picky about it.

hirenvadalia commented on src/cmds/scripts/pbs_habitat.in:
```diff
@@ -621,7 +269,7 @@ if [ "${PBS_START_SERVER:-0}" != 0 ] ; then
-
+	export PBS_DATA_SERVICE_USER
 	# create the database user file if it does not exist
 	if [ ! -f "${dbuser_fl}" ]; then
```
this if is useless here as `get_db_user` will create db user file if doesn't exist so at this point will always have db user file created

hirenvadalia commented on src/cmds/scripts/pbs_systemd:
```diff
@@ -0,0 +1,75 @@
+
+. ${PBS_CONF_FILE:-/etc/pbs.conf}
+
+is_systemd=0
```
I think we should rename this file to pbs_db_systemd, also shouldn't this file be part of Libdb/pgsql ? as this is specific to pgsql right?

hirenvadalia commented on src/include/pbs_db.h:
```diff
@@ -196,23 +155,23 @@ typedef struct pbs_db_job_info pbs_db_job_info_t;
-	BIGINT  ri_savetm;
+	INTEGER ri_fromsock;	/* resv from sock */
+	BIGINT  ri_fromaddr;	/* resv from sock addr */
+	BIGINT  ri_creattm;		/* resv create time on db */
```
please fix indentation in this file and remove extra spaces...

hirenvadalia commented on src/include/pbs_db.h:
```diff
@@ -505,14 +386,13 @@ int pbs_db_prepare_sqls(pbs_db_conn_t *conn);
  *
  */
-int pbs_db_begin_trx(pbs_db_conn_t *conn, int isolation_level, int async);
+int pbs_db_begin_trx(void *conn, int isolation_level);
```
We don't have trx db API, I suggest to move this in Libdb/pgsql

ashwathraop replied:
This should go when we remove trx code.

hirenvadalia commented on src/include/pbs_db.h:
```diff
@@ -959,64 +565,34 @@ int pbs_db_get_schema_version(pbs_db_conn_t *conn, int *db_maj_ver, int *db_min_
  *
  */
-char* pbs_db_get_svr_id(pbs_db_conn_t *conn, char *hostname);
+void get_db_errmsg(int err_code, char **err_msg);
```
can we rename this API to something like "pbs_db_get_errmsg"

hirenvadalia commented on src/include/pbs_db.h:
```diff
@@ -959,64 +565,34 @@ int pbs_db_get_schema_version(pbs_db_conn_t *conn, int *db_maj_ver, int *db_min_
  *
  */
-void pbs_db_reset_obj(pbs_db_obj_info_t *obj);
+int pbs_db_password(void *conn, char *userid, char *password, char *olduser);
```
can we rename this API to pbs_db_set_cred? or pbs_db_set_owner?

ashwathraop replied:
It matches the interface we use to change database password. So why change?

hirenvadalia commented on src/include/pbs_nodes.h:
```diff
@@ -54,7 +54,7 @@ extern "C" {
 #ifndef PBS_MOM
 #include "pbs_db.h"
-extern pbs_db_conn_t	*svr_db_conn;
+extern void	*svr_db_conn;
```
remove extra spaces between void and *

hirenvadalia commented on src/lib/Libdb/Makefile.am:
```diff
@@ -36,21 +36,11 @@
-	db_postgres_node.c \
-	db_postgres_sched.c
+clean-local:
+	rm -f $(top_builddir)/src/lib/Libdb/libdb.so
```
missing new line

hirenvadalia commented on src/lib/Libdb/pgsql/Makefile.am:
```diff
@@ -0,0 +1,68 @@
+	@database_lib@
+
+libdb_la_SOURCES = \
+	db_postgres.h \
```
we don't have to add header file in sources list

ashwathraop replied:
Sanitize build would fail without this.

hirenvadalia commented on src/lib/Libdb/pgsql/Makefile.am:
```diff
@@ -0,0 +1,68 @@
+install-exec-hook:
+	cp -p $(DESTDIR)$(libdir)/libdb.so.0.0.0 $(DESTDIR)$(libdir)/libdb_pg.so
+uninstall:
+	rm -f $(DESTDIR)$(libdir)/libdb_pg.so
```
missing new line

hirenvadalia commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1523 @@
+ * @file    db_common.c
+ *
+ * @brief
+ *      This file contains Postgres specific implementation of functions
```
indentation fix

hirenvadalia commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1523 @@
+ * An array of structures(of function pointers) for each of the database object
+ */
+pg_db_fn_t db_fn_arr[PBS_DB_NUM_TYPES] =
+	{
```
styling fix

hirenvadalia commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1523 @@
+ * @brief
+ *	Initialize a query state variable, before being used in a cursor
+ *
+ * @param[in]	conn - Database connection handle
```
missing query_cb in param[in]

hirenvadalia commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1523 @@
+ *
+ * @param[in]	conn - Database connection handle
+ *
+ * @return	Pointer to opaque cursor state handle
```
@return should say a type of return value instead an explanation of it

ashwathraop replied:
The opaque pointer is the type, right?

hirenvadalia commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1523 @@
+ *	the state variable
+ *
+ * @param[in]	st - Pointer to the state variable
+ *
```
missing @return

ashwathraop replied:
There is nothing to return. What to write in that case?

hirenvadalia commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1523 @@
+{
+	db_query_state_t *state = st;
+	if (state)
+	{
```
styling fix

hirenvadalia commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1523 @@
+	int ret;
+	int totcount = 0;
+	int rc;
+//	void *cb_ret = NULL;
```
remove commented line

nithinj commented on src/lib/Libdb/pgsql/pbs_db_utility.sh:
```diff
@@ -0,0 +1,1292 @@
+	exit 0
+}
+
+if [ "${opt}" = "upgrade_db" ]; then
```
upgrade and install functionalities were separated into two different files previously. Can we achieve the same here if there isn't much common code?

ashwathraop replied:
Yes, now I made them as functions within dbutility file. I wanted to reduce the number of database-specific files.

nithinj replied:
Separating functionalities into different files make it easier to maintain them.

ashwathraop replied:
You will not have these files when you move to another db. It will be an overhead to move back to pgsql since you have to copy multiple files now. With this only files you need to worry about are libdb.so and this script file.

nithinj replied:
I think we will be shipping PBS only one database at a time.
And if the customer wants to switch to a new database on the fly (could not find a use-case tho), still you will be ending up copying only these two as upgrade is required only for a schema change.

ashwathraop replied:
Both install and upgrade were part of install_db before too and install_db used to take an argument "upgrade" from pbs_habitat. And it was just some schema code was moved a different file.  

I am still not convinced that moving these 160 odd lines to a different file would add significant advantages to maintainability.

nithinj replied:
I realize that it was removed from pbs_habitat earlier. It should be fine as it resided in a database specific file now.

nithinj commented on src/server/queue_recov_db.c:
```diff
@@ -234,17 +239,17 @@ que_recov_db(char *qname)
 		goto db_err;
 
-	pbs_db_reset_obj(&obj);
+	free_db_attr_list(&(obj.pbs_db_un.pbs_db_que->attr_list));
```
you don't need an extra parenthesis here and in other places as & has less precedence over . and ->

ashwathraop replied:
Well, maybe for better readability and it doesnt harm to keep them is it?

nithinj replied:
I would have agreed if its a convoluted expression. But otherwise following the standard makes it easier for others to read the code.

ashwathraop commented on src/cmds/scripts/pbs_dataservice:
```diff
@@ -82,45 +82,34 @@ get_port() {
-
 CWD=`pwd`
 
 if [ ! -d "${PBS_HOME}/datastore/base" ]; then
```
Why?

ashwathraop commented on src/cmds/scripts/pbs_dataservice:
```diff
@@ -138,32 +127,32 @@ if [ "$myid" != "0" ]; then
+# Check if DBUSER is enabled, try to cd to user home
 if [ -n "$NASMODE" ] ; then
-	su - ${PGUSR} -s /bin/sh -c "cd" >/dev/null 2>&1
+	su - ${DBUSER} -s /bin/sh -c "cd" >/dev/null 2>&1
```
Dont we need NASMODE anymore?

ashwathraop commented on src/cmds/pbs_ds_password.c:
```diff
@@ -548,49 +541,19 @@ main(int argc, char *argv[])
 		/* change password only if this config option is not set */
 
-		if (pbs_db_begin_trx(conn, 0, 0) != 0) {
+		if (pbs_db_begin_trx(conn, 0) != 0) {
```
Eventually, trx code will go.

nithinj commented on src/server/pbsd_init.c:
```diff
@@ -2116,3 +2015,204 @@ call_log_license(struct work_task *ptask)
+ * @return 	NULL - on failure
+ */
+static job *
+refresh_job(pbs_db_obj_info_t *dbobj, int *totcount)
```
A better place for refresh functions would be the corresponding obj_recov_db.

ashwathraop replied:
But its being called only from pbsd_init. So do you think it makes sense to move it?

nithinj replied:
pbsd_init should consist of daemon initialization routines. whereas refresh_obj is associated with recovering the corresponding object. It is used in more places with multi-server.

ashwathraop replied:
I looked at multiserver team work branch, I mostly saw single use of these functions. can you direct me to right place where these functions are used multiple times?

nithinj replied:
never mind, this PR: https://github.com/subhasisb/pbspro/pull/85/files#diff-00dd478f2039ec72de85f6299fd806c8R630
will put them in the right place.

nithinj replied:
It will be easier to resolve later if you can put the functions in obj_recov_db itself.

ashwathraop replied:
Moved to respective recov_db files.

hirenvadalia commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1542 @@
+
+	/* create unique filename by appending pid */
+	sprintf(errfile, "%s/spool/db_errfile_%s_%d",
+		pbs_conf.pbs_home_path,
```
please put this in one line

ashwathraop commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1542 @@
+		query_cb(obj, &totcount);
+
+	pbs_destroy_state(st);
+	if (totcount > 0)
```
When query/search returns no results we return 1. For example, the server creates a default scheduler when it doesn't find one in the DB.

ashwathraop commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1542 @@
+	db_query_state_t *state = (db_query_state_t *)st;
+	int ret;
+
+	if (state->row < state->count) {
```
This is not an API function. This is internal and only used by LIbDB.

ashwathraop commented on src/cmds/pbs_dataservice.c:
```diff
@@ -0,0 +1,137 @@
+
+/**
+ * @brief
+ *	The main function in C - entry point
```
Removed.

ashwathraop commented on src/cmds/pbs_ds_password.c:
```diff
@@ -146,12 +147,13 @@ cleanup()
+
 	if (started_db  == 1) {
-		if (pbs_shutdown_db(&db_err) != 0) {
+		if (pbs_stop_db(conn_db_host, pbs_conf.pbs_data_service_port) != 0) {
```
The value of conn_db_host is set after the check.

ashwathraop commented on src/cmds/pbs_ds_password.c:
```diff
@@ -460,44 +460,43 @@ main(int argc, char *argv[])
-		db_conn_error = pbs_db_connect(conn);
-		if (db_conn_error == PBS_DB_SUCCESS && change_user == 1) {
+
+		failcode = pbs_db_connect(&conn, NULL, pbs_conf.pbs_data_service_port, PBS_DB_CNT_TIMEOUT_NORMAL);
```
No, you dont need to. When passed NULL, LIBDB assumes request is coming from localhost. See previous usage of pbs_db_init_connection.

ashwathraop commented on src/cmds/scripts/pbs_dataservice:
```diff
@@ -176,172 +158,63 @@ fi
 	if [ $? -ne 0 ]; then
 		echo "Could not change directory to ${PBS_HOME}/datastore"
-		exit 1
+		exit -1
```
Because it is an error case.

ashwathraop commented on src/include/pbs_db.h:
```diff
@@ -191,17 +151,17 @@ typedef struct pbs_db_que_info pbs_db_que_info_t;
-	char    nd_savetm[DB_TIMESTAMP_LEN + 1];
-	pbs_db_attr_list_t cache_attr_list; /* list of attributes */
-	pbs_db_attr_list_t db_attr_list; /* list of attributes */
+	char nd_name[PBS_MAXSERVERNAME+1];	/* vnode's name */
```
We have TEXT for char *.

ashwathraop commented on src/include/pbs_db.h:
```diff
@@ -308,35 +269,37 @@ struct pbs_db_query_options {
-#define PBS_DB_DOWN					1
-#define PBS_DB_STARTING				2
-#define PBS_DB_STARTED				3
+#define PBS_DB_DOWN		1
```
Looks alright to me. It might be the git diff. Please open the file in view mode and you can see indentation is okay.

ashwathraop commented on src/include/pbs_db.h:
```diff
@@ -352,140 +315,60 @@ typedef struct pbs_db_query_options pbs_db_query_options_t;
- *
- */
-int pbs_db_connect_async(pbs_db_conn_t *conn);
+int pbs_db_connect(void **conn, char * host, int port, int timeout);
```
To return a pointer you need **.

ashwathraop commented on src/include/pbs_db.h:
```diff
@@ -737,24 +529,7 @@ int pbs_startup_db_async(char **errmsg);
- *
- */
-int pbs_shutdown_db_async(char **errmsg);
+int pbs_stop_db(char *pbs_ds_host, int pbs_ds_port);
```
pbs_db.h is the interface file. LibDB header file will have a database specific data.

ashwathraop commented on src/include/svrfunc.h:
```diff
@@ -126,13 +127,14 @@ extern void spool_filename(job *, char *, char *);
+extern pbs_queue *find_resvqueuebyname(char *);
+extern resc_resv *find_resv_by_quename(char *);
+extern int make_schedselect(attribute *, resource *, pbs_queue *, attribute *);
+extern void free_db_attr_list(pbs_db_attr_list_t *);
```
If you mean free_db_attr_list(), its used by only server to work with attr_list attributes.

hirenvadalia commented on src/lib/Libdb/pgsql/pbs_db_utility:
```diff
@@ -0,0 +1,1272 @@
+
+generate_schema_file () {
+	# Create the schema file.
+	echo "/*" >  ${schema}
```
instead of echo per line, using `cat >file<__EOF__` might give more readable version of this function

ashwathraop replied:
Done.

hirenvadalia commented on src/lib/Libdb/pgsql/pbs_db_utility:
```diff
@@ -0,0 +1,1272 @@
+			echo "It appears that PostgreSQL has been upgraded independently of PBS Pro."
+			echo "The PBS Pro database must be manually upgraded. Please refer to the"
+			echo "instructions in the release notes for version @PBS_VERSION@ located here:"
+			echo "https://github.com/PBSPro/pbspro/releases"
```
just noting here: this message is changed in mainline and removed ref to url, so after rebase to mainline you may have to manually change it, or you can change now also if you want

ashwathraop replied:
Done.

ashwathraop commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1544 @@
+ * value if network and host byte order are identical.
+ */
+unsigned long long
+pbs_ntohll(unsigned long long x)
```
I dont want to link libutil just for this within libdb. So I will keep it this way.

nithinj commented on src/server/req_stat.c:
```diff
@@ -319,52 +286,40 @@ get_all_db_queues()
-	pbs_db_cursor_close(conn, cur_state);
-
 	if (count > 0) {
 		sprintf(log_buffer, "Refreshed %d queues", count);
```
Why is this printed in error level? This looks like a debug info. Applicable to other objects as well.

ashwathraop replied:
It was existing code but yes make sense to change it to DEBUG.

ashwathraop commented on src/server/req_stat.c:
```diff
@@ -128,53 +128,42 @@ get_all_db_nodes(char *hostname)
 {
```
Okay.

subhasisb commented on src/server/job_recov_db.c:
```diff
@@ -637,20 +653,20 @@ resv_recov_db(char *resvid, resc_resv *presv)
 	job *pj = NULL;
 	int load_type = 0;
-	*refreshed = 0;
+	pbs_db_job_info_t *dbjob = dbobj->pbs_db_un.pbs_db_job;
```
you need to set refreshed to 0, else in one path it may not get touched at all. And if the caller forgets to initialize it, then we are caput

subhasisb commented on src/server/job_recov_db.c:
```diff
@@ -670,6 +686,9 @@ refresh_job(pbs_db_job_info_t *dbjob, int *refreshed)
 		*refreshed = 1;
 	}
 
+	free_db_attr_list(&dbjob->db_attr_list);
```
dont u need free_db_xxx_list() calls in err case also? Looks like u have that in refresh_resv() but not in refresh_job()

subhasisb commented on src/server/job_recov_db.c:
```diff
@@ -678,27 +697,103 @@ refresh_job(pbs_db_job_info_t *dbjob, int *refreshed)
+ * @return 	NULL - on failure
+ */
+resc_resv *
+recov_resv_cb(pbs_db_obj_info_t *dbobj, int *refreshed)
```
Now that i think about it a bit more. I think it is bad and duplicate code to have both refresh_resv() and recov_resv_cb(). 

I think it function should be called resv_init() and refresh_resv() should call it each time it loads a reservation. This function should not do a recov_db anyway. It should simply deal with what to do when a reservation is loaded from disk (either during init or during stat for multiserver).

See, how i call pbsd_init_job() from refresh_job(). 

Btw, the better names for refresh_job() and refresh_resv() etc are job_recov_cb() resv_recov_cb() etc (the word refresh applies only for stats but we are using these functions for non-stat as well)

subhasisb commented on src/server/job_recov_db.c:
```diff
@@ -729,9 +824,15 @@ refresh_resv(pbs_db_resv_info_t *dbresv, int *refreshed)
 		*refreshed = 1;
 	}
 
+	free_db_attr_list(&dbresv->db_attr_list);
```
can we remove this duplicate code free_db_xxx_list() twice (once in good path, once in error path) - u can use an rc to return proper value

subhasisb commented on src/server/job_recov_db.c:
```diff
@@ -678,27 +697,103 @@ refresh_job(pbs_db_job_info_t *dbjob, int *refreshed)
 	char *at;
-
-	*refreshed = 0;
+	pbs_db_resv_info_t *dbresv = dbobj->pbs_db_un.pbs_db_resv;
```
u need to initialize refreshed

subhasisb commented on src/server/job_recov_db.c:
```diff
@@ -729,12 +826,117 @@ refresh_resv(pbs_db_resv_info_t *dbresv, int *refreshed)
+	}
+
+	/* to save the last job's time save_tm, since we are loading in order */
+	if (strncmp(opts.timestamp, jobs_from_time, DB_TIMESTAMP_LEN) > 0)
```
u dont need to compare - u can directly assign it

subhasisb commented on src/lib/Libdb/pgsql/db_common.c:
```diff
@@ -0,0 +1,1484 @@
+			strcpy(opts->timestamp,((pbs_db_resv_info_t *) obj->pbs_db_un.pbs_db_resv)->ri_savetm);
+			break;
+		case PBS_DB_QUEUE:
+			strcpy(opts->timestamp, ((pbs_db_que_info_t *) obj->pbs_db_un.pbs_db_que)->qu_savetm);
```
for completeness, you should add cases for sched and server objects as well (even though today we do not search them)

subhasisb commented on src/server/pbsd_main.c:
```diff
@@ -1259,144 +1231,23 @@ main(int argc, char **argv)
+	/* determine the rate of calling the ping_nodes functionality based on server_init_type */
+	if (server_init_type == RECOV_HOT) {
+			/* rapid ping rate while hot restart */
+			ping_nodes_rate = HOT_START_PING_RATE < svr_ping_rate ? HOT_START_PING_RATE : svr_ping_rate;
```
we don't have pings any more from server, so is ping_nodes_rate required? please check with @nithinj

nithinj replied:
It's not used any longer. I'm removing it as part of my mainline PR. So you can either remove or keep it for the time being.

subhasisb commented on src/server/node_recov_db.c:
```diff
@@ -463,15 +464,16 @@ refresh_node(pbs_db_obj_info_t *dbobj, int *refreshed)
 	free_db_attr_list(&dbnode->db_attr_list);
```
Can we not have 2 sets of free_db_attr_list() in these functions?

subhasisb replied:
i i think u just need to delete this pair

ashwathraop commented:
@subhasisb I addressed your comments. I did not touch refresh_nodes at this time since it looks drastically different than init version. I also talked to @nithinj and he says init version will change soon.

ashwathraop commented:
![image](https://user-images.githubusercontent.com/17821341/83463582-74a22d80-a43c-11ea-99c1-b9938de0fb32.png)

Here are the results from TH3. Failures compare to integration branch.

bremanandjk commented:
FYI: Now integration branch is capable to run simple job run tests in multi-server mode. Please try to run it once. Thanks

ashwathraop commented:
@bremanandjk do you have steps or automated scripts for this?

bremanandjk commented:
@ashwathraop You could get more details about multi-server switches in this doc. https://pbspro.atlassian.net/wiki/spaces/PD/pages/1585545237/Sharding+algorithm+design

For automation, as we know, it is decided to do it later.
Meanwhile, I can share quick steps.

Add these 2 configs in pbs.conf
1) Set PBS_MAX_SERVERS = <num of servers >
2) Set PBS_SERVER_INSTANCES = [[host]:port,][[host]:port,]...
3) Start the servers using a pbs startup script and submit some jobs. In your case, plz make sure that it is not breaking any existing functionalities.

ashwathraop commented:
[test_results.txt](https://github.com/subhasisb/openpbs/files/4719954/test_results.txt)
Here is the test result for multi-server test.

ashwathraop commented:
![image](https://user-images.githubusercontent.com/17821341/83588968-9248c300-a520-11ea-8229-4eebfb7be540.png)

@subhasisb I pushed changes please take a look. On TH3 failures are not consistent so it is little hard to make which was existing and which is not. I am still looking at this.

subhasisb commented on src/server/job_recov_db.c:
```diff
@@ -816,28 +767,38 @@ refresh_resv(pbs_db_obj_info_t *dbobj, int *refreshed)
-		
-		*refreshed = 1;
 
+		is_resv_window_in_future(presv);
```
can we put this in a function call like pbsd_init_resv()

ashwathraop commented:
@subhasisb Moved the code to pbsd_init_resv() function. Please take a look.

ashwathraop commented:
![image](https://user-images.githubusercontent.com/17821341/83659344-64e82d80-a591-11ea-82b6-4157c513dbd1.png)
![image](https://user-images.githubusercontent.com/17821341/83659369-6b76a500-a591-11ea-9762-b718120be5dc.png)
TH3 results.

ashwathraop commented:
![image](https://user-images.githubusercontent.com/17821341/83659602-bc869900-a591-11ea-870e-e673dac02d49.png)
TH3 on integration branch to compare.
